### ISSUE TITLE
LuaJit support

### ISSUE THREAD

**Issue by manast** [dragonflydb/dragonfly#2043]
Currently, Dragonfly uses Lua 5.4.4 This is the latest version of LUA, and I think it is nice to be able to support newer versions in comparison with stock Redis which has not updated the LUA engine since 5.1.

However, for more CPU-intensive LUA scripts (as the ones used for example in BullMQ), it may be beneficial to use [LuaJIT](https://luajit.org/luajit.html) instead. The tradeoff is that it is based on an older Lua syntax (5.1), but the performance could be several times faster.

I am not sure if LuaJIT is going to be beneficial or not, nor how much work it would be to implement a POC to do some benchmarks, but it would be nice to know for sure if it is something that is worth having or on the other hand the benefits are minimal and thus discard this possibility.

**Comment by kostasrim** [comment]
Hi @manast , I am not quire sure what kind of performance increase we are looking to with `LuaJIT` here. Is that something that `redis` support as well? Do you have some data on how much we can increase the performance by using `Jit` ? 

My two cents here are if this does not contribute to a huge performance boost then it's not really worth our time since there probably are other optimizations that can offer a better performance gain for the effort. 

I am not an expert on this feature so maybe @chakaz or @romange can add a little bit more info here as well. 

From a quick google search I stumbled on this article:

https://staff.fnwi.uva.nl/h.vandermeer/docs/lua/luajit/luajit_performance.html#:~:text=The%20ratio%20between%20the%20times,or%20with%20%2Dj%20off). 

it seems like there is a nice boost, but does that mean that it's relevant for bullmq scripts? :shrug:

**Comment by manast** [comment]
@kostasrim In some benchmarks it is up to 10x faster than Lua, but as I wrote, I have no idea how performance would be increased in practice, that's why I am suggesting if a quick and dirty integration is possible then we could see if it is worth to have a production-ready integration or not.

**Comment by kostasrim** [comment]
@manast I like the idea. let's wait for the rest of the team and their opinion. 

Thank you for bringing this up :)

**Comment by chakaz** [comment]
Hi @manast and thanks for the suggestion!
Roman and I discussed this today, and we believe that currently there are lower hanging fruits to tackle before trying this out, but perhaps in the future we'll give this one a shot as well.
FWIW Roman saw that most of the time spent in Lua is related to memory allocations, which will likely remain the same with any implementation.

**Comment by chakaz** [comment]
Today I ran a cool experiment, which was Roman's idea:
I hard-coded addJob's commands so that when addJob is called on my server, it runs the exact commands that the script would have run but run them directly without using Lua.
This gave us an upper limit for how much in theory we could improve Lua (which in practice would obviously be much less).
The results are pretty much statistical noise: I can see improvement of <2%, which is something but again we can't remote _all_ overhead.
As such, we decided not to pursue this, at least not at the moment.

We do have some pretty cool optimizations that we're working on, unrelated to Lua directly, which should be included in the next release (:crossed_fingers:)
